### PR TITLE
feat: add boringcrypto builds for linux

### LINKED ISSUE #9087: Build Coder with boringcrypto
Google has a fork of BoringSSL that is designed to meet the FIPS spec. While Google is non-commital about this [^1], using boringcrypto is an important step when some enterprise users evaluate Coder. 

[^1]: https://github.com/golang/go/blob/master/src/crypto/internal/boring/README.md

### PR DESCRIPTION
**Pull request by spikecurtis**
fixes #9087 

Adds support for building with Boring Crypto on Linux, when not cross-compiling.

It's maybe a little surprising for us to build different versions of Coder binaries depending on where you build.  I chose to do it this way so that everyone gets boring crypto if available on their platform, so that we're not having people develop against non-boring-crypto and then shipping a different build that we haven't been testing with.

~It also leaves debug symbols in boring crypto binaries, since it is hard to verify we are building with boring crypto without these.  **This increases binary sizes** by about 5MiB for the linux amd64 slim binary (39 -> 44) and 8MiB (184 -> 192) for the fat linux amd64 binary.~

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -105,6 +105,17 @@ CODER_ARCH_IMAGE_PREREQUISITES := \
...
+ifeq ($(OS),Windows_NT)
+	local_os:=Windows
+	local_arch:="" #ignored, no boringcrypto support for Windows
+else
+	local_os:=$(shell uname -s)
+	local_arch:=$(shell uname -m)
+endif
+ifeq ($(local_arch),x86_64)
+	local_arch:=amd64
+endif
```
Comment (deansheather) [inline Makefile:118]: You might be able to simplify this with `shell go env GOHOSTOS` and `GOHOSTARCH`

File: scripts/build_go.sh [hunk 2]
```diff
@@ -94,11 +102,16 @@ if [[ "$sign_darwin" == 1 ]]; then
 fi
 
 ldflags=(
-	-s
 	-w
 	-X "'github.com/coder/coder/v2/buildinfo.tag=$version'"
 )
 
+# For boringcrypto we want to leave the symbols so we can verify it was build correctly for
+# FIPS compliance.  This adds a few MiB to the binary.
```
Comment (johnstcn) [inline scripts/build_go.sh:110]:
```suggestion
# For boringcrypto we want to leave the symbols so we can verify it was built correctly.
# This adds a few MiB to the binary.
```

**Review by deansheather — APPROVED** [review]

**Review by johnstcn — APPROVED** [review]
Do we need to _build_ with boringcrypto, or just provide the _option_ to build with boringcrypto? It's a shame that we need to leave the debugging symbols in.

**Comment by mafredri** [comment]
Re: debug symbols, could we strip them after we’ve verified the build?

File: Makefile (hunk 1 above)
Comment (spikecurtis) [reply Makefile:118]: Oh! That's much nicer

**Comment by spikecurtis** [comment]
> Re: debug symbols, could we strip them after we’ve verified the build?

We could, but then customers wouldn't be able to verify.

**Comment by spikecurtis** [comment]
> Do we need to _build_ with boringcrypto, or just provide the _option_ to build with boringcrypto? It's a shame that we need to leave the debugging symbols in.

We need to be able to ship boringcrypto builds, and I don't think we should ship variants for a single OS/arch: i.e. if you run on Linux amd64 you get boringcrypto.

**Comment by mafredri** [comment]
> We could, but then customers wouldn't be able to verify.

I was thinking we would add `ldflags+=(-X "'github.com/coder/coder/v2/buildinfo.boringcrypto=true'")` and output this in `coder version [--output json]`.

We could also archive the non-stripped binaries for potential verification and/or use where symbols maybe be needed.

**Comment by mafredri** [comment]
> I was thinking we would add `ldflags+=(-X "'github.com/coder/coder/v2/buildinfo.boringcrypto=true'")` and output this in `coder version [--output json]`.
> 
> We could also archive the non-stripped binaries for potential verification and/or use where symbols maybe be needed.

Actually, on second thought, we should utilize build tags here `//go:build boring` => set true and let the compiler take care of it.

**Comment by spikecurtis** [comment]
Thanks @mafredri --- added to BuildInfo

File: scripts/build_go.sh [hunk 3]
```diff
@@ -140,7 +148,15 @@ cmd_path="./enterprise/cmd/coder"
...
-CGO_ENABLED=0 GOOS="$os" GOARCH="$arch" GOARM="$arm_version" go build \
+
+cgo=0
+goexp=""
+if [[ "$boringcrypto" == 1 ]]; then
+	cgo=1
+	goexp="boringcrypto"
+fi
+
+GOEXPERIMENT="$goexp" CGO_ENABLED="$cgo" GOOS="$os" GOARCH="$arch" GOARM="$arm_version" go build \
```
Comment (mafredri) [inline scripts/build_go.sh:159]: Does `-boringcrypto` also need to be added to `go build` args?

**Review by mafredri — APPROVED** [review]
Looking nice!

Edit: Looks like `TestVersion/JSON_output` needs updating.

**Pull request merged**